File: Source/Q_Method/Q_Method.cpp
#include "Q_Method.h"

Eigen::MatrixXd B_solve(
	Eigen::MatrixXd b1,
	Eigen::MatrixXd b2,
	Eigen::MatrixXd b3,
	Eigen::MatrixXd r1,
	Eigen::MatrixXd r2) {
	/****************************************
	Takes components of sun vectors and magnetic field vectors
	and returns B matrix using weights w1, w2, and w3.
	input: b1, b2, b3, r1, r2
	output: B
	****************************************/

	int w1 = 1, w2 = 1, w3 = 1;
	Eigen::MatrixXd B;
	B = w1 * (b1 * r1.conjugate().transpose()) + w2 * (b2 *
		r2.conjugate().transpose()) + w3 * (b3 *
		r2.conjugate().transpose());
	return B;
}

Eigen::MatrixXd S_solve(Eigen::MatrixXd B) {
	/****************************************
	Takes B matrix and returns S matrix.
	input: B
	output: S
	****************************************/

	Eigen::MatrixXd S;
	S = B + B.conjugate().transpose();
	return S;
}

double sigma_solve(Eigen::MatrixXd B) {
	/****************************************
	Takes B matrix and returns sigma coefficient.
	input: B
	output: sigma
	****************************************/

	double sigma = B.trace();
	return sigma;
}
Eigen::MatrixXd Z_solve(Eigen::MatrixXd B) {
	/****************************************
	Takes B matrix and returns Z matrix.
	input: B
	output: Z
	****************************************/

	Eigen::MatrixXd Z(3, 1), Z_initialize(1, 3);
	Z_initialize << B(1, 2) - B(2, 1), B(2, 0) - B(0, 2), B(0, 1) - B(1, 0);
	Z = Z_initialize.conjugate().transpose();
	return Z;
}

Eigen::MatrixXd K_solve(
	Eigen::MatrixXd b1,
	Eigen::MatrixXd b2,
	Eigen::MatrixXd b3,
	Eigen::MatrixXd r1,
	Eigen::MatrixXd r2) {
	/****************************************
	Takes components of sun vectors and magnetic field vectors
	and returns K matrix.
	input: b1, b2, b3, r1, r2
	output: K
	****************************************/

	double sigma;
	Eigen::MatrixXd K(4, 4), B, S, Z;
	B = B_solve(b1, b2, b3, r1, r2);
	S = S_solve(B);
	sigma = sigma_solve(B);
	Z = Z_solve(B);

	/*
	   * K is a 4x4 matrix of the form:
	   * | S-σI   Z|
	   * |	       |
	   * | ZT     σ|
	   */
	   //Initializes S-σI section of K matrix.
	K(0, 0) = S(0, 0) - sigma * 1;
	K(0, 1) = S(0, 1) - sigma * 0;
	K(0, 2) = S(0, 2) - sigma * 0;
	K(1, 0) = S(1, 0) - sigma * 0;
	K(1, 1) = S(1, 1) - sigma * 1;
	K(1, 2) = S(1, 2) - sigma * 0;
	K(2, 0) = S(2, 0) - sigma * 0;
	K(2, 1) = S(2, 1) - sigma * 0;
	K(2, 2) = S(2, 2) - sigma * 1;

	//Initializes Z section of K matrix.
	K(0, 3) = Z(0, 0);
	K(1, 3) = Z(1, 0);
	K(2, 3) = Z(2, 0);

	//Initializes Z conjugate transpose section of K matrix.
	K(3, 0) = Z.conjugate().transpose()(0, 0);
	K(3, 1) = Z.conjugate().transpose()(0, 1);
	K(3, 2) = Z.conjugate().transpose()(0, 2);

	//Initializes σ section of K matrix.
	K(3, 3) = sigma;
	return K;
}

Eigen::MatrixXd quaternion_estimate(Eigen::MatrixXd K) {
	/****************************************
	Takes K matrix and returns quaternion estimate matrix.
	input: B
	output: q_est
	****************************************/

	int index;
	Eigen::MatrixXd V, D, q_est;
	Eigen::EigenSolver<Eigen::MatrixXd> eigenSolver(K);

	//Initializes V, a matrix with columns of eigenvectors from K matrix.
	V = (eigenSolver.eigenvectors()).real();

	//Initializes D, a matrix of eigenvalues from K matrix.
	D = (eigenSolver.eigenvalues()).real();

	//Finds the index of the maximum eigenvalue.
	for (int i = 0; i < D.rows(); i++) {
		if (D(i, 0) == D.maxCoeff()) {
			index = i;
		}
	}

	//Sets quaternion estimate to be the eigenvector corresponding to the index.
	q_est = V.col(index);
	return q_est;
}

Eigen::MatrixXd q_method(
	Eigen::MatrixXd b1,
	Eigen::MatrixXd b2,
	Eigen::MatrixXd b3,
	Eigen::MatrixXd r1,
	Eigen::MatrixXd r2) {
	/****************************************
	Takes components of sun vectors and magnetic field vectors
	and returns quaternion estimate matrix.
	input: b1, b2, b3, r1, r2
	output: q_est
	****************************************/

	Eigen::MatrixXd K, q_est;
	K = K_solve(b1, b2, b3, r1, r2);
	q_est = quaternion_estimate(K);
	return q_est;
}

File: Matlab/Q_Method/q_Method.m
function q_est = q_Method(b1,b2,b3,r1,r2)
% q-Method Attitude Determination Algorithm
% DEFINE INPUTS/OUTPUTS
% Input:
% ======
% uses two different vectors (eg. Sun and magnetic field) who's components
% are expressed in Body-Fixed and ECI frames:
% b1 = components of Sun vector expressed in Body-Fixed frame (from Diodes)
% b2 = components of magnetic field vector A expressed in Body-Fixed frame
%      (from magnetometer A)
% b3 = components of magnetic field vector B expressed in Body-Fixed frame
%      (from magnetometer B)
% r1 = components of Sun vector expressed in ECI frame
%      (from theory)
% r2 = components of magnetic field vector expressed in ECI frame
%      (from theory)
% NOTE: the above are column matrices
%       ie:  b1=[b1x,b1y,b1z]'
%
% Output:
% =======
%   q_est  = estimated quaternion orientation

% reference: Markley, F., Crassidis, J., 
% "Fundmentals of Spacecraft Attitude Determination and Control"
% Space Technology Library, Springer, 2014, page 184
%
% also based on Example 4.3 in Chris Hall's March 18 2003 Ch4 notes
% online: http://www.dept.aoe.vt.edu/~cdhall/courses/aoe4140/attde.pdf
%
% This software is for Dal CubeSat project internal use only and
% is a work-in-progress.
% This software is provided "as is" and any expressed or implied 
% warrantees, including, but not limited to, the implied warranties
% of merchantability and fitness for a particular purpose are disclaimed.
% Dr. Robert Bauer shall not be liable for any direct, indirect,
% consequential, or other damages suffered by anyone resulting from this
% work or the use of the research results/data of this work.
%%
w1=1; % weightimg matrix for first measurement (Sun vector)
w2=1; % weightimg matrix for second measurement (mag field vector A)
w3=1; % weightimg matrix for third measurement (mag field vector B)

B = w1*(b1*r1')+w2*(b2*r2')+w3*(b3*r2'); %if have two mag field measurements
S = B + B';
Z = [B(2,3)-B(3,2), B(3,1)-B(1,3), B(1,2)-B(2,1)]';
sigma = trace(B);

% q-Method:
% based on Example 4.3 in Chris Hall's March 18 2003 notes
% online: http://www.dept.aoe.vt.edu/~cdhall/courses/aoe4140/attde.pdf
% also summarized in Fundamentals of Spacecraft Attitude Determination
% on page 187, (Eq references refer to this text)
K = [S-sigma*eye(3,3),     Z;
                   Z', sigma]; %Eq 5.17
 
% the eigenvector corresponding to the largest eigenvalue of K is the
% least-squares optimal estimate of the attitude
[V,D]=eig(K); %calculate diagonal matrix D of eigenvalues and
              %matrix V whos columns are corresponding eigenvectors
D=diag(D);    %convert diagonal matrix D into a column vector of eigenvalues
[value,indx]=max(real(D)); %find index of the maximum eigenvalue
q_est = real(V(:,indx)); % use corresponding eigenvector as the 
                         % estimated attitude quaternion 
                         % in [xi yj zk w] format, Eq. 5.23
                
                         

